### PR TITLE
Add cost and enforcer task

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
Cost and enforcer task process

     * 1. make require properties for children base of current operator
     * 2. try to optimize child and get best (Cost, GroupPlan) pair of every children
     * 3. once we get all output properties of one candidate loop, derive output properties base of current operator
     * 4. if output properties does not satisfied require properties, add enforcers and submit (Cost, GroupPlan) pair

### CODE REVIEW

File: src/property/mod.rs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+use crate::operator::Operator;
+use crate::property::sort_property::SortProperty;
+use crate::Plan;
+
+pub trait Property {}
+pub trait LogicalProperty: Property {}
+pub trait PhysicalProperty: Property {
+    fn satisfy(&self, _input: PhysicalProperties) -> bool
+    where
```
Comment (linbo-lin) [inline src/property/mod.rs]: Why Self:Sized

File: src/task.rs [hunk 2]
```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
...
+        // let plan = output_prop.order_property.add_enforcer();
+    }
+
+    fn prev_index(&self) -> usize {
+        self.prev_index
+    }
+
+    fn submit_cost_plan(&self, child_output_props: &[Rc<PhysicalProperties>]) {
+        let mut curr_plan = &self.plan;
+        let output_prop = curr_plan.borrow().operator().derive_output_prop(child_output_props);
```
Comment (linbo-lin) [inline src/task.rs]: 1.let output_prop = curr_plan.borrow().operator().physical_op().derive_output_prop(child_output_props);
2.then you can delete Operator::derive_output_prop()

File: src/task.rs [hunk 3]
```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
...
+    /**
+     * 1. make require property for children base of current operator
+     * 2. try to optimize child group and get best (Cost, GroupPlan) pair of every children
+     * 3. once we get all output property of one candidate loop, derive output property base of current operator
+     * 4. if output property does not satisfied require property, add enforcers and submit (Cost, GroupPlan) pair
+     */
+    fn execute(mut self, task_runner: &mut TaskRunner, _optimizer_ctx: &mut OptimizerContext) {
+        // 1. according to current operator create new requestPropList for children
+        let reqd_props_list = self.make_required_props_list();
+        for (index, required_props) in reqd_props_list.iter().enumerate() {
```
Comment (linbo-lin) [inline src/task.rs]: 1. for (index, required_props) in reqd_props_list.iter().enumerate().skip(self.prev_index)
2. then delete 
```
if index < self.prev_index() {
    continue;
}
```

File: src/property/mod.rs (hunk 1 above)
Comment (LiBinfeng-01) [reply src/property/mod.rs]: IDE told me to do that...let me try to remove this

File: src/task.rs (hunk 2 above)
Comment (LiBinfeng-01) [reply src/task.rs]: get

File: src/task.rs (hunk 3 above)
Comment (LiBinfeng-01) [reply src/task.rs]: get

File: src/task.rs [hunk 4]
```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
...
+            let mut child_output_props = Vec::new();
+            if index < self.prev_index() {
+                continue;
+            }
+            for (required_prop, child_group) in required_props.iter().zip(self.plan.borrow().inputs()) {
+                // 2. optimize children groups using requestPropList
+                match child_group.borrow().lowest_cost_plans().get(required_prop) {
+                    Some((_cost, plan)) => {
+                        // add life time parameter let output life time equals to self
```
Comment (linbo-lin) [inline src/task.rs]: Delete useless comment

File: src/task.rs [hunk 5]
```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
...
+        let mut curr_plan = &self.plan;
+        let output_prop = curr_plan.borrow().operator().derive_output_prop(child_output_props);
+
+        let mut curr_cost = curr_plan.borrow().compute_cost();
+        if !output_prop.satisfy(&self.required_prop) {
+            // add enforcer
+            curr_plan = self.add_enforcer_to_group(&output_prop);
+            curr_cost = curr_plan.borrow().compute_cost();
+        }
+        match curr_plan
```
Comment (linbo-lin) [inline src/task.rs]: let curr_group = curr_plan.borrow().group();
match curr_group.borrow().lowest_cost_plans().get(&self.required_prop) {...}
curr_group.borrow_mut().lowest_cost_plans_mut().insert(self.required_prop.clone(), (curr_cost, (*curr_plan).clone()));

File: src/task.rs [hunk 6]
```diff
@@ -59,23 +61,36 @@ impl TaskRunner {
 
 pub struct OptimizeGroupTask {
     group: GroupRef,
+    required_prop: PhysicalProperties,
```
Comment (linbo-lin) [inline src/task.rs]: `Rc<PhysicalProperties>`

File: src/task.rs [hunk 7]
```diff
@@ -130,16 +145,111 @@ impl OptimizePlanTask {
...
+                }
+            }
+        }
+        // update lowest_cost_plans
+        curr_plan
+            .borrow_mut()
+            .group()
+            .borrow_mut()
+            .lowest_cost_plans_mut()
+            .insert(self.required_prop.clone(), (curr_cost, (*curr_plan).clone()));
```
Comment (linbo-lin) [inline src/task.rs]: `(curr_cost, curr_plan.clone())`

File: src/operator/mod.rs [hunk 8]
```diff
@@ -7,9 +7,11 @@ pub mod physical_scan;
 pub mod physical_topn;
 
 use crate::metadata::MdAccessor;
+use crate::property::PhysicalProperties;
 use crate::statistics::Statistics;
 use std::any::Any;
 use std::rc::Rc;
+use std::todo;
```
Comment (linbo-lin) [inline src/operator/mod.rs]: can be deleted, and compilations can also pass

File: src/memo.rs [hunk 9]
```diff
@@ -85,19 +99,21 @@ pub struct Group {
     physical_plans: Vec<GroupPlanRef>,
     is_explored: bool,
     statistics: Option<Rc<Statistics>>,
+    lowest_cost_plans: HashMap<Rc<PhysicalProperties>, (Cost, GroupPlanRef)>,
```
Comment (linbo-lin) [inline src/memo.rs]: you can `type LowestCostPlans = HashMap<Rc<PhysicalProperties>, (Cost, GroupPlanRef)>;
` and use LowestCostPlans

File: src/memo.rs [hunk 10]
```diff
@@ -1,16 +1,20 @@
...
 use std::cell::RefCell;
+use std::collections::HashMap;
 use std::rc::{Rc, Weak};
 
 pub struct GroupPlan {
     group: GroupWeakRef,
     op: Operator,
     inputs: Vec<GroupRef>,
     rule_masks: BitSet,
+    require_to_output_map: HashMap<PhysicalProperties, Rc<PhysicalProperties>>,
```
Comment (linbo-lin) [inline src/memo.rs]: you can `type RequireToOutputMap = HashMap<PhysicalProperties, Rc<PhysicalProperties>>;` and use RequireToOutputMap

Comment (LiBinfeng-01) [reply src/memo.rs]: done

File: src/memo.rs (hunk 9 above)
Comment (LiBinfeng-01) [reply src/memo.rs]: done

File: src/operator/mod.rs (hunk 8 above)
Comment (LiBinfeng-01) [reply src/operator/mod.rs]: done

File: src/task.rs (hunk 7 above)
Comment (LiBinfeng-01) [reply src/task.rs]: done

File: src/task.rs (hunk 6 above)
Comment (LiBinfeng-01) [reply src/task.rs]: done

File: src/memo.rs [hunk 11]
```diff
@@ -77,27 +84,40 @@ impl GroupPlan {
         let input_stats = input_stats.as_slice();
         self.op.logical_op().derive_statistics(md_accessor, input_stats)
     }
+
+    pub fn get_output_prop(&self, reqd_prop: &PhysicalProperties) -> &Rc<PhysicalProperties> {
```
Comment (davidli2010) [inline src/memo.rs:88]: get_output_properties

File: src/memo.rs [hunk 12]
```diff
@@ -77,27 +84,40 @@ impl GroupPlan {
...
+        self.require_to_output_map.get(reqd_prop).expect("output not null")
```
Comment (davidli2010) [inline src/memo.rs:89]: Is `expect` safe?

File: src/cost/mod.rs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+#[derive(PartialEq, PartialOrd)]
+pub struct Cost {
```
Comment (davidli2010) [inline src/cost/mod.rs]:
```rust
#[repr(transparent)]
pub struct Cost(f64);
```

**Review by davidli2010 — CHANGES_REQUESTED** [review]

**Review by linbo-lin — APPROVED** [review]

**Pull request merged**